askama_derive/src/generator.rs

```diff
@@ -727,6 +727,57 @@ impl<'a> Generator<'a> {
                 args.len()
             )));
         }
+        // Handling both named and unnamed arguments requires to be careful of the arguments order.
+        // To do so, we iterate through the macro defined arguments and then check if we have a
+        // named argument with this name:
+        //
+        // * If there is one, we add it and move to the next argument.
+        // * If there isn't one, we then need to pick the next unnamed argument (as we consider they
+        //   are passed in the right order). To ensure we won't accidentally pass a named argument,
+        //   we split named and unnamed arguments in two vectors before starting.
+        let mut output_args = Vec::new();
+        // Since named arguments can only be passed last, we only need to check if the last argument
+        // is a named one.
+        let args = if args
+            .last()
+            .map(|arg| matches!(arg, Expr::NamedArgument(_, _)))
+            .unwrap_or(false)
+        {
+            // We have named arguments, so we need to ensure that they are passed in the right
+            // order.
+            //
+            // To make things easier, we split them in two different vectors so we won't need to
+            // check each time we need to push an unnamed argument if it isn't, in fact, a named
+            // one.
+            let (mut named_arguments, mut non_named_arguments): (Vec<_>, Vec<_>) = args
+                .iter()
+                .partition(|item| matches!(item, Expr::NamedArgument(_, _)));
+
+            // First we check that all named arguments actually exist in the called item.
+            for arg in named_arguments.iter() {
```

Comment:
I think the right logic here is:

- Set up a `named_args` `HashMap<&str, &Expr>`
- For each argument, in reverse order:
  - If the argument is not named, break out of the loop
  - If it is named, insert into `named_args`
- Then, change the `iter::zip()` loop to add `.enumerate()`
  - If an argument is named, use the `named_args` map to lookup the actual argument